File: ja2/src/chipset/Memory.java
package chipset;

import java.io.InputStream;
import java.util.Arrays;
import java.util.Random;
import util.Util;




/*
 * Created on Aug 1, 2007
 */
public class Memory
{
	private static final byte CLEAR_VALUE = 0;

	private final byte[] bytes;

	/**
	 * @param keyboard
	 */
	public Memory(final int bytes)
	{
		this.bytes = new byte[bytes];
	}

	public int size()
	{
		return this.bytes.length;
	}

	public byte read(final int address)
	{
		return this.bytes[address];
	}

	public void write(final int address, final byte data)
	{
		this.bytes[address] = data;
	}

	public void load(int base, final InputStream in) throws InvalidMemoryLoad
	{
		try
		{
			for (int byt = in.read(); byt != Util.EOF && base < this.bytes.length; byt = in.read())
			{
				this.bytes[base++] = (byte)byt;
			}
		}
		catch (final Throwable e)
		{
			throw new InvalidMemoryLoad(e);
		}
	}

	public void clear()
	{
		Arrays.fill(this.bytes,CLEAR_VALUE);
	}
}

File: epple2/memory.cpp
#include "memory.h"
#include <vector>
#include <algorithm>
#include <istream>

static const int CLEAR_VALUE(0);

Memory::Memory(const size_t n):
	bytes(n)
{
}

size_t Memory::size() const
{
	return this->bytes.size();
}

unsigned char Memory::read(const unsigned short address) const
{
	return this->bytes[address];
}

void Memory::write(const unsigned short address, const unsigned char data)
{
	this->bytes[address] = data;
}

void Memory::clear()
{
	std::fill(this->bytes.begin(),this->bytes.end(),CLEAR_VALUE);
}

void Memory::powerOn()
{
//	const RAMInitializer initRam(*this);
//	initRam.init();
}

void Memory::powerOff()
{
	clear();
}

void Memory::load(std::istream& in)
{
	in.read((char*)this->bytes.front(),this->bytes.size());
}
